### PR TITLE
win: patch from OSGeo4W applied

### PR DESCRIPTION
**Pull request by landam**
This PR applies https://github.com/jef-n/OSGeo4W/blob/master/src/grass/osgeo4w/patch (author: @jef-n)

### CODE REVIEW

**Comment by neteler** [comment]
A dependency yet missing?

From the CI log:
```
...
configure: error: *** couldn't find libpng-config
Error: Process completed with exit code 1.
```

**Comment by nilason** [comment]
> A dependency yet missing?
> 
> From the CI log:
> 
> ```
> ...
> configure: error: *** couldn't find libpng-config
> Error: Process completed with exit code 1.
> ```

Not a dependency, but https://github.com/OSGeo/grass/blob/main/mswindows/osgeo4w/libpng-config (see #2679).

**Comment by jef-n** [comment]
Note that build_osgeo4w.sh and osgeo4w.yml are GRASS things, that is not used in osgeo4w - they were just aligned with package.sh, which actually is.

**Comment by landam** [comment]
After reintroducing [mingw-w64-x86_64-libpng](https://github.com/OSGeo/grass/pull/4121/commits/daba1e6c55b815d609264af52c9a76905f9eecce) the CI pass successfully. 

@jef-n Please could you elaborate why you are removing this dependency in the workflow (https://github.com/jef-n/OSGeo4W/blob/master/src/grass/osgeo4w/patch#L9)? On the other side it's installed in `package.sh` (https://github.com/jef-n/OSGeo4W/blob/master/src/grass/osgeo4w/package.sh#L101)

**Comment by jef-n** [comment]
> After reintroducing [mingw-w64-x86_64-libpng](https://github.com/OSGeo/grass/pull/4121/commits/daba1e6c55b815d609264af52c9a76905f9eecce) the CI pass successfully.
> 
> @jef-n Please could you elaborate why you are removing this dependency in the workflow (https://github.com/jef-n/OSGeo4W/blob/master/src/grass/osgeo4w/patch#L9)? On the other side it's installed in `package.sh` (https://github.com/jef-n/OSGeo4W/blob/master/src/grass/osgeo4w/package.sh#L101)

Because osgeo4w has libpng.

**Comment by jef-n** [comment]
There was a change in libpng that broke our copy of libpng-config - adapted.

**Comment by landam** [comment]
Any idea why osgeo4w CI is failing on tests while compilation is successful? It seems to be unrelated to this PR...

**Comment by echoix** [comment]
There were a lot more failed files, like 58% of files passed, instead of 81%

**Comment by landam** [comment]
> Any idea why osgeo4w CI is failing on tests while compilation is successful? It seems to be unrelated to this PR...

Solved by [b435503](https://github.com/OSGeo/grass/pull/4121/commits/b43550307c568120c0150a63abb370804c9179a7)

**Comment by echoix** [comment]
There seems to be a bad rebase somewhere, there are some unrelated commits in the PR.

Was there a logic in the new order of the OSGeo4W dependencies, was it in the patch that it was in that order? It's not alphabetical neither, I don't find the logic. I'll try to compare manually which ones are changed, as word-based diff highlighting doesn't show the changes clearly.

**Review by nilason — COMMENTED** [review]
Looks good to me in general. I have some thoughts on one change, which I'd like to look into though.

File: include/Make/Platform.make.in
Comment (nilason) [reply include/Make/Platform.make.in:237]: Just tested successfully locally without this change. There should be no need to add `OPENMP_CFLAGS` to `OPENMP_LIB`.

File: mswindows/osgeo4w/package.sh
Comment (hellik) [inline mswindows/osgeo4w/package.sh:1]: should we document somewhere why and which DLL is needed to include in the winGRASS package? IIRC I was involved in a trial/error-way to check which of the DLLs we added finally back then.

File: mswindows/osgeo4w/package.sh
Comment (hellik) [inline mswindows/osgeo4w/package.sh:245]: this should be checked whether we include these files already in another way, e.g. in the NSIS installer-

File: mswindows/osgeo4w/env.bat.tmpl
Comment (hellik) [inline mswindows/osgeo4w/env.bat.tmpl:12]: what does this mean?

**Review by hellik — COMMENTED** [review]
sse my inline comments for a final check

**Comment by nilason** [comment]
> Was there a logic in the new order of the OSGeo4W dependencies, was it in the patch that it was in that order? It's not alphabetical neither, I don't find the logic. I'll try to compare manually which ones are changed, as word-based diff highlighting doesn't show the changes clearly.

The current changes are:

```diff
 cairo-devel
-fftw
 freetype-devel
 gdal-devel
-gdal-ecw
-gdal-mrsid
 geos-devel
+libjpeg-turbo-devel
 liblas-devel
 libpng-devel
 libpq-devel
 libtiff-devel
-libxdr
 netcdf-devel
-pdal-devel
-pdcurses
 proj-devel
-python3-matplotlib
+python3-core
 python3-numpy
 python3-ply
 python3-pywin32
+python3-six
 python3-wxpython
-regex-devel
+sqlite3-devel
 zstd-devel
```

**Comment by echoix** [comment]
Thanks @nilason! I don't have a strong opinion on the changed dependencies, apart from matplotlib that I thought it was included as a dependency in other builds, but I don't know if it is really used in our modules or just in addons...

File: include/Make/Platform.make.in
Comment (echoix) [reply include/Make/Platform.make.in:237]: On the contrary, I tried using the package.sh build on CI, more closer to OSGeo4W build, and I had some gomp errors: https://github.com/echoix/grass/actions/runs/10641753457/job/29503311365?pr=207#step:9:1864

**Comment by echoix** [comment]
I fixed the conflicts, and took the new deps and sorted them into the action (it would have worked without the multiline list too, but now the diff is clear)

**Comment by echoix** [comment]
Fixed conflicts

**Comment by nilason** [comment]
osgeo4w installs libomp.dll:
https://github.com/jef-n/OSGeo4W/blob/f8ada2fe6637b1abde0a361fa8b5b2089a82e33d/src/grass/osgeo4w/package.sh#L104

which is probably why configure hooks up on it.

Perhaps 

```diff
-/mingw64/bin/libgomp-1.dll
+/mingw64/bin/libomp.dll
```
in https://github.com/OSGeo/grass/blob/3f61cd7625f728ff0795023e97529e2ec896c5cc/mswindows/osgeo4w/package.sh#L127
could be the solution?

File: .github/workflows/osgeo4w.yml
Comment (echoix) [inline .github/workflows/osgeo4w.yml:66]: Since we have some tests in the grass.benchmark (python/grass/benchmark/testsuite/)using matplotlib, should we be keeping it installed?
```suggestion
            python3-matplotlib
            python3-numpy
```

Comment (hellik) [reply .github/workflows/osgeo4w.yml:66]: Yes, matplotlib should be there

**Comment by nilason** [comment]
At least 
OpenMP issues need to be addressed for 8.4.1, I’ll add backport label for this PR as not to miss this.

**Comment by echoix** [comment]
The conflicts in the PR I might have time this weekend to address them, it's not too long, it's the sorted lines that need to be chosen with respect to the changes this PR made

**Comment by echoix** [comment]
I solved the conflicts by sorting the two conflicting files of the PR in same way as https://github.com/OSGeo/grass/pull/4563, then comparing the two sides of the merge making sure changes present in the PR (on GitHub) were still present even if moved. It was all ok, nothing is lost.

**Comment by echoix** [comment]
> osgeo4w installs libomp.dll:
> https://github.com/jef-n/OSGeo4W/blob/f8ada2fe6637b1abde0a361fa8b5b2089a82e33d/src/grass/osgeo4w/package.sh#L104
> 
> which is probably why configure hooks up on it.
> 
> Perhaps 
> ```diff
> -/mingw64/bin/libgomp-1.dll
> +/mingw64/bin/libomp.dll
> ```
> in https://github.com/OSGeo/grass/blob/3f61cd7625f728ff0795023e97529e2ec896c5cc/mswindows/osgeo4w/package.sh#L127
> could be the solution?



I've reread on a fresh mind. Isn't libomp the llvm OpenMP runtime library (https://packages.msys2.org/packages/mingw-w64-x86_64-openmp, files section at the very end of https://packages.msys2.org/packages/mingw-w64-x86_64-llvm-openmp)
 
And libgomp the gcc OpenMP library
(https://packages.msys2.org/packages/mingw-w64-x86_64-omp, files section at the very end of https://packages.msys2.org/packages/mingw-w64-x86_64-gcc-libs)

(And libiomp the intel OpenMP one, not in msys2).


So, following the often linked advice  that multiple OpenMP runtimes shouldn't be mixed (https://cpufun.substack.com/p/is-mixing-openmp-runtimes-safe), and that we are using the gcc toolchain, I think we should be removing all references to libomp.dll, and go all-in with libgomp-1.dll, and work until fixing all the issues.

**Comment by nilason** [comment]
OpenMP related fix (hopefully) is added with https://github.com/jef-n/OSGeo4W/commit/06d4d83912dffccd638417338f43f03f4f84c15b.

**Comment by echoix** [comment]
Conflicts to solve

**Comment by nilason** [comment]
OpenBLAS issue for main has been addressed with https://github.com/OSGeo/grass/commit/9984205dc8eaab7f014a5117374565a479297d9d, which doesn't need to backport.

OpenMP issues have been addressed (as mentioned above), but doesn't need changes in GRASS source (and thus no need to backport).

I removed backport label for this PR, as the build issues have likely been addressed.

**Comment by nilason** [comment]
@landam Could you please check if the win build server is working, as there has been no build log since end of August.

**Comment by echoix** [comment]
I'm still processing the changed patch 16 hours ago in the OSGeo4W repo

**Comment by echoix** [comment]
In the last commit I'm keeping (for now) the `--with-libpng=${SRC}/mswindows/osgeo4w/libpng-config \` line that was removed. I think it might me a merging/rebase conflict that lost it, as I would have expected that line to be changed to use `${SRC}` instead of `${PWD}` like the other places in the file that got updated for that. If it is expected, then we should also remove it from our parallel script.

File: include/Make/Platform.make.in
Comment (echoix) [reply include/Make/Platform.make.in]: This was dropped upstream as we fixed the OpenMP problem with #4716, https://github.com/jef-n/OSGeo4W/pull/24, https://github.com/jef-n/OSGeo4W/pull/25. Resolving.

File: mswindows/osgeo4w/env.bat.tmpl
Comment (echoix) [reply mswindows/osgeo4w/env.bat.tmpl:12]: The first one, PYTHONPATH, is needed, I often had to add it to make windows tests work in CI (like when working hard to get pytest running), but the second one I don't know

File: mswindows/osgeo4w/package.sh
Comment (echoix) [reply mswindows/osgeo4w/package.sh:1]: We could add a separate issue for this, as we're mostly only applying the patches that have been used upstream. Something quite simple that we've spent 4 months on, so I would hope to not stretch it further if no easy consensus comes.

File: mswindows/osgeo4w/package.sh
Comment (echoix) [reply mswindows/osgeo4w/package.sh:246]: Is this exact script really used for the NSIS installer? This script is used only by OSGeo4W (yet) from the best of my knowledge. I have some work prepared even before this PR was filed to use this script for our CI, to be really closer to what OSGeo4W packages, and could upload the artifacts that could be tested locally, and use ccache (that is already scripted here, we only have to download and upload the cache in CI). To not duplicate work, I was patiently waiting for this PR to be merged before continuing, but I would really want to continue to allow shaving some time off of the longest CI job that we have.

**Review by echoix — COMMENTED** [review]
I've gone through the remaining conversations to resolve. Like I mention in these, I would really like to have this finished, so I could resume some other work that I rebased on this (as a couple weeks before this PR I already did a similar effort and continued on this), but I've been waiting 4 months since. Recently we solved one of the blockers, so there's not much left to think here.

File: mswindows/osgeo4w/build_osgeo4w.sh
Comment (echoix) [inline mswindows/osgeo4w/build_osgeo4w.sh:49]: See https://github.com/OSGeo/grass/pull/4121#issuecomment-2495506323:

> In the last commit I'm keeping (for now) the `--with-libpng=${SRC}/mswindows/osgeo4w/libpng-config \` line that was removed. I think it might me a merging/rebase conflict that lost it, as I would have expected that line to be changed to use `${SRC}` instead of `${PWD}` like the other places in the file that got updated for that. If it is expected, then we should also remove it from our parallel script.

File: mswindows/osgeo4w/env.bat.tmpl
Comment (jef-n) [reply mswindows/osgeo4w/env.bat.tmpl:12]: the fatal version check in gisinit.c that used to be patched out in osgeo4w now just causes a warning if GRASS_COMPATIBILITY_CHECK is 0.

File: mswindows/osgeo4w/package.sh
Comment (jef-n) [inline mswindows/osgeo4w/package.sh:244]: if this is merged, DLLS can be removed

File: mswindows/osgeo4w/env.bat.tmpl
Comment (echoix) [reply mswindows/osgeo4w/env.bat.tmpl:12]: Oh, so you're talking about this part:
https://github.com/OSGeo/grass/blob/3e22b87722fb96c86c27f3e78405bc8c4fe0ed54/lib/gis/gisinit.c#L43-L73

Shouldn't it be `GRASS_COMPATIBILITY_TEST`?

File: mswindows/osgeo4w/package.sh
Comment (jef-n) [reply mswindows/osgeo4w/package.sh:246]: No, NSIS does that here:
https://github.com/OSGeo/grass/blob/3e22b87722fb96c86c27f3e78405bc8c4fe0ed54/mswindows/GRASS-Packager.bat.tmpl#L65

File: mswindows/osgeo4w/package.sh
Comment (echoix) [reply mswindows/osgeo4w/package.sh:1]: > a libpng update in osgeo4w broke "our" version of `libpng-config` and probably made someone switch from osgeo4w's libpng to mingw's mingw-w64-x86_64-libpng along with it's default `libpng-config` to fix the build.

For reference, this comment seems to be a reply for https://github.com/OSGeo/grass/pull/4121/files/2031013418a44cadf855e4add44673945c7925bb#r1855246849

File: mswindows/osgeo4w/env.bat.tmpl
Comment (echoix) [reply mswindows/osgeo4w/env.bat.tmpl:12]: > Shouldn't it be `GRASS_COMPATIBILITY_TEST`?

The comment by Jürgen was edited to match this

Comment (echoix) [reply mswindows/osgeo4w/env.bat.tmpl:12]: Thus, this explained what that change was, and this conversation will be resolved

File: mswindows/osgeo4w/build_osgeo4w.sh
Comment (echoix) [reply mswindows/osgeo4w/build_osgeo4w.sh:49]: I'm choosing to keep for now, so that the file mswindows/osgeo4w/libpng-config changed in this PR would be used. https://github.com/OSGeo/grass/pull/4121/files#diff-67926bb21aac5cfc85772a5dd151278e4743e182391cd750abc02c299c87f97c

File: mswindows/osgeo4w/package.sh
Comment (echoix) [reply mswindows/osgeo4w/package.sh:1]: > should we document somewhere why and which DLL is needed to include in the winGRASS package? IIRC I was involved in a trial/error-way to check which of the DLLs we added finally back then.

This comment: https://github.com/OSGeo/grass/pull/4121/files#r1855518495 seems to say that this list might not be needed soon. So I won't request more work on that for now

File: mswindows/osgeo4w/package.sh
Comment (echoix) [reply mswindows/osgeo4w/package.sh:223]: I checked, and indeed, with these lines, the $DLLS variable isn't used anymore

**Review by echoix — APPROVED** [review]
I did a final review, and all conversations are addressed.

**Comment by hellik** [comment]
Thanks for checking all these improvements.

**Comment by echoix** [comment]
@jef-n The daily grass-dev builds seems to have stopped after this got merged, no new builds were made since december 15, 2024. I wanted to let you know

**Pull request merged**
